src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,157 @@
 """Implementation of Rule L003."""
-from typing import List, Optional, Sequence, Tuple
+import dataclasses
+import itertools
+from typing import Dict, List, Optional, Sequence, Set, Tuple
 
 from sqlfluff.core.parser import WhitespaceSegment
-from sqlfluff.core.parser.segments import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments import BaseSegment
 from sqlfluff.core.rules.functional import rsp, sp, Segments
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import (
     document_fix_compatible,
     document_configuration,
 )
 from sqlfluff.core.templaters import TemplatedFile
+from sqlfluff.core.templaters.base import RawFileSlice
+
+
+@dataclasses.dataclass
+class _LineSummary:
+    """A dataobject to represent a line.
+
+    A _LineSummary is created and then filled with elements,
+    before calling self.from_current_line to generate a final
+    representation.
+    """
+
+    line_no: int = 0
+    line_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    # As of end of line
+    indent_balance: int = 0
+    # As it was as of the "Anchor" / first code elem
+    anchor_indent_balance: int = 0
+    line_anchor: Optional[BaseSegment] = None
+
+    # Fixed calculated values
+    templated_line: Optional[int] = None
+    hanging_indent: Optional[int] = None
+    indent_size: int = 1
+    clean_indent: bool = True
+    templated_line_type: Optional[str] = None
+    is_comment_line: bool = False
+    is_empty_line: bool = False
+    has_code_segment: bool = False
+
+    line_indent_stack: List[int] = dataclasses.field(default_factory=list)
+    hanger_pos: Optional[int] = None
+
+    def __repr__(self) -> str:
+        """Printable Summary without Segments."""
+        keys_to_strip = (
+            "line_buffer",
+            "indent_buffer",
+            "as_of_anchor",
+        )
+        print_dict: Dict = {
+            key: value
+            for key, value in self.__dict__.copy().items()
+            if key not in keys_to_strip
+        }
+        print_dict["raw_line"] = self.template_content
+        return print_dict.__repr__()
+
+    @property
+    def template_content(self):  # pragma: no cover
+        return "".join(
+            seg.raw or getattr(seg, "source_str", "") for seg in self.line_buffer
+        )
+
+    def from_current_line(self, line_no: int, templated_file: Optional[TemplatedFile]):
```

Comment:
I went with `finalise` and `working_state` for the variable. I feel that makes it clear as to your point.

I actually tried for quite a while to figure a pattern with two distinct classes where one is frozen, but anything I tried added significant overhead which just felt like it was making it more complicated.

cause _WorkingLineState needs to be able to create _LineSummary and vis-vera and also linting complains about inheriting and freezing (I think it would actually function at runtime)